### PR TITLE
Scroll to top

### PR DESCRIPTION
**Pull request by LibbyChapman**
@incuna/frontend please review

Made floaty scroll to top button

This is the first bit of javascript/jquery for this project and I'm not sure if I've done the right things to add it in? Added jquery to the bower.json, do I need to do anything else? There are no tests and that seems strange :D

### CODE REVIEW

**Comment by hippogriffic** [comment]
@ninjanails set up some basic travis but we probably don't have any js testing yet, I will make an issue to add it but it would be helpful if somebody can tell me the process for that @wytrych @henrahmagix ?

File: scripts/scroll-to-top.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+        $(document).on('click', topButton, function () {
+            performScroll(scrollableAreaTop);
+        });
+
+        $(window).on('resize', throttle(function () {
+            setUpTopButton();
+        }, throttleInterval));
+    });
+}(window.jQuery, window._));
```
Comment (hippogriffic) [inline scripts/scroll-to-top.js]: wow so much javascripty

**Comment by hippogriffic** [comment]
@incuna/frontend issue for adding JS testing is here: https://incuna-pm.atlassian.net/browse/SL-33
please update description with something useful, or tell me what to put there, thanks

**Comment by LibbyChapman** [comment]
Thanks @hippogriffic. Sorry I'm a bit clueless with the project setup stuff, will do/learn what I can :)

**Comment by hippogriffic** [comment]
yeah no worries, I've been here almost 2 years and I still mostly have no idea how to do this stuff :D

File: scripts/scroll-to-top.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    var windowPosition;
+    var footerPosition;
+    var buttonPosition;
+    var scrollableAreaTopOffset;
+    var windowTop;
+    var showTopButtonDistance;
+    var throttleInterval = 100;
+
+    // Throttle function from https://remysharp.com/2010/07/21/throttling-function-calls
+    function throttle(fn, threshhold, scope) {
```
Comment (Minglee01) [inline scripts/scroll-to-top.js]: this should be `var throttle = function`
also it looks formatted wrong

File: scripts/scroll-to-top.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    var performScroll = function (target) {
+        $('html, body').animate({
+            scrollTop: $(target).offset().top
+        }, 'medium');
+        setUpTopButton();
+    };
+
+    $(document).ready(function () {
+
+        if (!$('body').hasClass('style-guide-page'))
```
Comment (Minglee01) [inline scripts/scroll-to-top.js]: should be 

```
 if (!$('body').hasClass('style-guide-page')) {
      return;
 }
```

File: scripts/scroll-to-top.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            return;
+        }
+        scrollableAreaTopOffset = $(scrollableAreaTop).offset().top;
```
Comment (Minglee01) [inline scripts/scroll-to-top.js:77]: Can potentially do  

```
scrollableAreaTop.offset().top
```

where scrollableAreaTop is

```
scrollableAreaTop = '$(.sg-jump-top');
```

File: scripts/scroll-to-top.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            var now = +new Date;
+            var args = arguments;
+            if (last && now < last + threshhold) {
+                // hold on to it
+                clearTimeout(deferTimer);
+                deferTimer = setTimeout(function () {
+                    last = now;
+                    fn.apply(context, args);
+                },
+                threshhold + last - now);
```
Comment (Minglee01) [inline scripts/scroll-to-top.js]: do you need the `)` ?

Comment (LibbyChapman) [inline scripts/scroll-to-top.js]: yeah, it closes the setTimeout(..

Comment (Minglee01) [inline scripts/scroll-to-top.js]: it looks wierd i think it should be 

```
 }, threshhold + last - now);
```

File: Gruntfile.js [hunk 6]
```diff
@@ -96,4 +113,17 @@ module.exports = function (grunt) {
...
+        'jscs'
+    ]);
+
+    grunt.registerTask('test', 'Run the tests.', function (env) {
+         'lint'
+    });
+
+    grunt.registerTask('travis', 'Run the tests in Travis', [
+        'test:travis'
+    ]);
```
Comment (henrahmagix) [inline Gruntfile.js]: I don't think you need this `travis` task, just `grunt test` in travis will do. This repo is unlikely to need to have separate local and travis tests

**Comment by henrahmagix** [comment]
This is 👍  I know the scroll script has been copied from a project, so I didn't review it heavily

**Comment by henrahmagix** [comment]
@LibbyChapman So have `- grunt test` in Travis, and delete the `travis` grunt task =)

**Comment by LibbyChapman** [comment]
thanks for the help everyone :)

**Comment by hippogriffic** [comment]
woo yay 🍰  🐐  ⭐  🐟

**Pull request merged**